### PR TITLE
Featureimage/ipmprog 1244 discovery migration

### PR DESCRIPTION
**Pull request by ericclappier-eaton**

### CODE REVIEW

**Comment by ericclappier-eaton** [comment]
Need https://github.com/42ity/fty-security-wallet/pull/62 for compilation

File: common/commands.h [hunk 1]
```diff
@@ -78,37 +78,185 @@ namespace commands::assets {
...
+            {
+                Unknown,
+                CancelledByUser,
+                Terminated,
+                InProgress
+            };
+
+            pack::Enum<Status> status     = FIELD("status");
+            pack::String       progress   = FIELD("progress");
+            pack::Int64        discovered = FIELD("discovered");
```
Comment (FrancoisRegisDegott-eaton) [inline common/commands.h:181]: these counters shall be unsigned (uint32_t or size_t)

File: common/commands.cpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    {
+        using Type = In::Type;
+
+        ss << [&]() {
+            switch (value) {
+                case Type::Local:
+                    return "localscan";
+                case Type::Ip:
+                    return "ipscan";
+                case Type::Multy:
```
Comment (FrancoisRegisDegott-eaton) [inline common/commands.cpp:16]: is it time to fix the typo? (Multi)

File: common/commands.h [hunk 3]
```diff
@@ -78,37 +78,185 @@ namespace commands::assets {
...
-        pack::String subAddress = FIELD("sub_address", "-1");
-        Asset        asset      = FIELD("asset");
+    } // namespace status
+
+    struct Response : public pack::Node
+    {
+        enum class Status
+        {
+            Success,
+            Fail,
```
Comment (FrancoisRegisDegott-eaton) [inline common/commands.h:201]: Failure

File: common/discovery-config-manager.h [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+#include <optional>
+#include <pack/pack.h>
+
+
+namespace fty::disco {
+class ConfigDiscoveryManager
+{
+public:
+    static ConfigDiscoveryManager& instance();
+    fty::Expected<void>            save(const std::string& path = ConfigFile);
```
Comment (FrancoisRegisDegott-eaton) [inline common/discovery-config-manager.h:14]: save() const?

File: common/discovery-config.cpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            case Type::Local:
+                return "localscan";
+            case Type::Ip:
+                return "ipscan";
+            case Type::Multy:
+                return "multiscan";
+            case Type::Full:
+                return "fullscan";
+            case Type::Unknown:
+                return "UNKNOWN";
```
Comment (FrancoisRegisDegott-eaton) [inline common/discovery-config.cpp:33]: why this 'unknown' is capitalized?

File: common/discovery-config.h [hunk 6]
```diff
@@ -0,0 +1,115 @@
+#pragma once
+#include <fty/expected.h>
+#include <pack/pack.h>
+
+namespace fty::disco {
+
+static constexpr const char* ConfigFile = "/etc/fty-discovery-ng/discovery.conf";
```
Comment (FrancoisRegisDegott-eaton) [inline common/discovery-config.h:7]: shall be ConfigFilePath (for readabitily)

File: common/discovery-config.h [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        pack::Int32  type = FIELD("type");
+
+        using pack::Node::Node;
+        META(DefaultValuesLink, src, type);
+    };
+
+    struct Parameters : public pack::Node
+    {
+        pack::String mappingFile       = FIELD("mappingFile");
+        pack::Int32  maxDumpPoolNumber = FIELD("maxDumpPoolNumber");
```
Comment (FrancoisRegisDegott-eaton) [inline common/discovery-config.h:77]: counters are unsigned...

File: common/discovery-task.h [hunk 8]
```diff
@@ -88,15 +88,14 @@ class Task : public fty::Task<T>
     {
     }
 
-    void operator()() override
+    Task() = default;
+
+    inline void operator()() override
```
Comment (FrancoisRegisDegott-eaton) [inline common/discovery-task.h:93]: for understanding and readability, some comments would be useful

File: common/message-bus.cpp [hunk 9]
```diff
@@ -30,13 +30,14 @@ namespace fty::disco {
...
+Expected<void> MessageBus::init(const std::string& actorName, const std::string& endpoint)
 {
     std::lock_guard<std::mutex> lock(m_mutex);
     try {
         m_bus = std::unique_ptr<messagebus::MessageBus>(messagebus::MlmMessageBus(endpoint, actorName));
         m_bus->connect();
         m_actorName = actorName;
+        m_endpoint = endpoint;
         return {};
     } catch (std::exception& ex) {
```
Comment (FrancoisRegisDegott-eaton) [inline common/message-bus.cpp:42]: const std::exception&

File: common/commands.cpp [hunk 10]
```diff
@@ -0,0 +1,123 @@
+#include "commands.h"
+
+namespace fty::disco {
```
Comment (FrancoisRegisDegott-eaton) [inline common/commands.cpp:3]: i like disco, but here i prefer discovery as a namespace

File: rest/conf/42_discovery.xml.in [hunk 11]
```diff
@@ -13,3 +13,45 @@
...
+  <method>GET</method>
+</mapping>
+
+<mapping>
+    <target>discovery/scan/stop@lib${NAME}</target>
+    <url>^/api/v1/admin/scan-ng.*$</url>
+    <method>DELETE</method>
+</mapping>
+
+<!-- TBD TO REMOVE -->
```
Comment (FrancoisRegisDegott-eaton) [inline rest/conf/42_discovery.xml.in:52]: is it time to remove that?

File: rest/CMakeLists.txt [hunk 12]
```diff
@@ -2,14 +2,29 @@
...
+        src/config-create.h
+        src/config-create.cpp
+        src/scan-start.h
+        src/scan-start.cpp
+        src/scan-stop.h
+        src/scan-stop.cpp
+        src/scan-status.h
+        src/scan-status.cpp
+        
+        # TBD TO REMOVE
```
Comment (FrancoisRegisDegott-eaton) [inline rest/CMakeLists.txt:24]: is it time to remove these files?

File: rest/src/discovery-rest.h [hunk 13]
```diff
@@ -0,0 +1,19 @@
+#pragma once
+#include "message.h"
+
+namespace fty::disco {
+
+static constexpr const char* AgentName = "discovery-ng_rest";
```
Comment (FrancoisRegisDegott-eaton) [inline rest/src/discovery-rest.h:6]: shall be fty-discovery-ng_rest

File: rest/src/discovery-rest.h [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+
+
+inline fty::disco::Message message(const std::string& subj)
+{
+    fty::disco::Message msg;
+    msg.meta.to      = "discovery-ng";
```
Comment (FrancoisRegisDegott-eaton) [inline rest/src/discovery-rest.h:12]: fty- prefix

File: rest/src/mibs.cpp [hunk 15]
```diff
@@ -60,7 +60,7 @@ Expected<std::string> Mibs::mibs(const commands::mibs::In& param)
     msg.meta.subject = commands::mibs::Subject;
     msg.meta.from    = "discovery_rest";
```
Comment (FrancoisRegisDegott-eaton) [inline rest/src/mibs.cpp:61]: use AgentName defined above (discovery-ng_rest)

File: rest/src/protocols.cpp [hunk 16]
```diff
@@ -61,7 +61,7 @@ Expected<std::string> Protocols::protocols(const commands::protocols::In& param)
     msg.meta.subject = commands::protocols::Subject;
     msg.meta.from    = "discovery_rest";
```
Comment (FrancoisRegisDegott-eaton) [inline rest/src/protocols.cpp:62]: AgentName

File: server/conf/discovery.conf.in [hunk 17]
```diff
@@ -1,3 +1,5 @@
 actor-name: 'discovery-ng'
```
Comment (FrancoisRegisDegott-eaton) [inline server/conf/discovery.conf.in:1]: missed fty- prefix?

File: server/conf/discovery.conf.in [hunk 18]
```diff
@@ -1,3 +1,5 @@
...
 log-config: 'logger.conf'
```
Comment (FrancoisRegisDegott-eaton) [inline server/conf/discovery.conf.in:2]: do we need a specific logger?

File: server/src/jobs/auto-discovery.h [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+    static const uint32_t SCAN_MIN_NUM_THREAD  = 5;
+
+    // TBD: To centralise with rest status ???
+    enum class State {
+        Unknown,
+        CancelledByUser,
+        Terminated,
+        InProgress
+    };
+    struct StatusDiscovery {
```
Comment (FrancoisRegisDegott-eaton) [inline server/src/jobs/auto-discovery.h:40]: shall be named Status (at least discoveryStatus)

File: server/src/jobs/auto-discovery.h [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+    // Construct and update output ext attributes according input ext attributes
+    static Expected<void> updateExt(const commands::assets::Ext& ext_in, asset::create::Ext& ext_out);
+
+    // Update host name
+    static Expected<void> updateHostName(const std::string& address, asset::create::Ext& ext);
+
+    // Device centric view
+    bool isDeviceCentricView() const
+    {
+        return (m_params.parent == "0") ? false : true;
```
Comment (FrancoisRegisDegott-eaton) [inline server/src/jobs/auto-discovery.h:83]: return (m_params.parent != "0");   //more readable

File: server/src/jobs/auto-discovery.h [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+    };
+
+    // Read configuration
+    Expected<void> readConfig(const disco::commands::scan::start::In& in);
+
+    // Status management
+    void statusDiscoveryInit();
```
Comment (FrancoisRegisDegott-eaton) [inline server/src/jobs/auto-discovery.h:90]: discoveryStatus instead...

File: server/src/jobs/create-asset.h [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+
+    using pack::Node::Node;
+    META(PowerLink, source, link_type, src_out);
+};
+
+class MapExtObj: public pack::Node
+{
+public:
+    pack::String value     = FIELD("value");
+    pack::Bool   readOnly  = FIELD("readOnly");
```
Comment (FrancoisRegisDegott-eaton) [inline server/src/jobs/create-asset.h:39]: seems CamelCase naming is not the rule here (read_only?)

File: server/src/jobs/impl/address.h [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+    static bool isRange(const std::string& range);
+
+    static fty::Expected<std::pair<std::string, std::string>> cidrToLimits(const std::string& cidr);
+    static fty::Expected<std::pair<std::string, std::string>> rangeToLimits(const std::string& range);
+
+    static fty::Expected<std::vector<std::string>> getRangeIp(const std::string& range);
+    static fty::Expected<std::vector<std::string>> getLocalRangeIp();
+
+private:
+    static fty::Expected<std::string> ntop(struct in_addr ip);
```
Comment (FrancoisRegisDegott-eaton) [inline server/src/jobs/impl/address.h:48]: const struct in_addr& ?

File: server/src/jobs/impl/address.h [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+    static int maskNbBit(std::string mask);
```
Comment (FrancoisRegisDegott-eaton) [inline server/src/jobs/impl/address.h:49]: const std::string&  (for mask and address)

File: server/src/jobs/impl/nut/process.cpp [hunk 25]
```diff
@@ -117,7 +118,8 @@ Expected<void> Process::init(const std::string& address, uint16_t port)
 
 Expected<std::string> Process::findExecutable(const std::string& name) const
 {
-    static std::vector<std::filesystem::path> paths = {"/usr/lib/nut", "/lib/nut", "/home/jes/workspace/fty/build/Debug/deps-runtime/bin"};
+    static std::vector<std::filesystem::path> paths = {
+        "/usr/lib/nut", "/lib/nut", "/home/jes/workspace/fty/build/Debug/deps-runtime/bin"};
```
Comment (FrancoisRegisDegott-eaton) [inline server/src/jobs/impl/nut/process.cpp:122]: /lib/nut only

File: server/src/jobs/impl/xml-pdc.cpp [hunk 26]
```diff
@@ -34,11 +34,11 @@ std::optional<std::string> Properties::value(const std::string& key)
 
 // =====================================================================================================================
 
-XmlPdc::XmlPdc(const std::string& address)
-    : m_ne(address)
+XmlPdc::XmlPdc(const std::string& address, const uint32_t port)
```
Comment (FrancoisRegisDegott-eaton) [inline server/src/jobs/impl/xml-pdc.cpp:37]: const port... useful?

File: test/protocols.cpp [hunk 27]
```diff
@@ -1,97 +1,164 @@
...
+namespace fty::disco {
+
+TEST_CASE("Protocols/ Empty request", "[protocols]")
 {
-    fty::disco::Message msg = Test::createMessage(fty::commands::protocols::Subject);
+    fty::disco::Message msg = Test::createMessage(commands::protocols::Subject);
 
     fty::Expected<fty::disco::Message> ret = Test::send(msg);
     CHECK_FALSE(ret);
     CHECK("Wrong input data: payload is empty" == ret.error());
```
Comment (FrancoisRegisDegott-eaton) [inline test/protocols.cpp:13]: so, if someone change/update the error() message... it break the tests. 
Why? Is it an external requirement for that?

File: test/test-common.h [hunk 28]
```diff
@@ -9,71 +9,32 @@
 #include <catch2/catch.hpp>
 #include <fty_log.h>
 #include <thread>
-
-
-class Test;
-inline Test* inst;
+#include <mlm_library.h>
```
Comment (FrancoisRegisDegott-eaton) [inline test/test-common.h:12]: #include <malamute.h> will suffice (and includes czmq.h)

**Pull request closed without merging**